File: ios/KarhooSdk.m
#import <React/RCTBridge.h>

@interface RCT_EXTERN_MODULE(KarhooSdk, NSObject)
    // Initialize
    //
    RCT_EXTERN_METHOD(initialize:(NSString *)identifier referer:(NSString *)referer organisationId:(NSString *)organisationId isProduction:(BOOL)isProduction)
    
    // KarhooUserService
    //
    RCT_EXTERN_METHOD(register:(NSDictionary *)registrationData resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)
    RCT_EXTERN_METHOD(login:(NSDictionary *)loginData resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)
    RCT_EXTERN_METHOD(logout)
    RCT_EXTERN_METHOD(currentUser)

    // KarhooAuthService
    //
    RCT_EXTERN_METHOD(loginWithToken:(NSString *)token resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)

    // KarhooAddressService
    //
    RCT_EXTERN_METHOD(placeSearch:(NSDictionary *)placeSearchData resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)
    RCT_EXTERN_METHOD(locationInfo:(NSDictionary *)locationInfoData resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)

    // KarhooQuotesService
    //

    // KarhooPaymentService
    //
    RCT_EXTERN_METHOD(getPaymentNonce:(NSDictionary *)paymentData resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)

    // KarhooTripService    
    //
    RCT_EXTERN_METHOD(bookTrip:(NSDictionary *)bookTripData resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)
    RCT_EXTERN_METHOD(cancellationFee:(NSString *)followCode resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)
    RCT_EXTERN_METHOD(cancelTrip:(NSString *)followCode resolve:(RCTPromiseResolveBlock *)resolve reject:(RCTPromiseRejectBlock *)reject)
@end

File: ios/KarhooSdk.swift
import KarhooSDK
import BraintreeDropIn

struct KarhooConfiguration: KarhooSDKConfiguration {
    var identifier: String
    var referer: String
    var organisationId: String
    var isProduction: Bool

    init (identifier: String, referer: String, organisationId: String, isProduction: Bool) {
        self.identifier = identifier
        self.referer = referer
        self.organisationId = organisationId
        self.isProduction = isProduction
    }

    func environment() -> KarhooEnvironment {
        return self.isProduction ? .production : .sandbox
    }

    func authenticationMethod() -> AuthenticationMethod {
        return .guest(settings: GuestSettings(identifier: self.identifier, referer: self.referer, organisationId: self.organisationId))
    }
}

@objc(KarhooSdk)
class KarhooSdk: NSObject { 
    @objc func initialize(_ identifier: String, referer: String, organisationId: String, isProduction: Bool) -> Void {
        Karhoo.set(configuration: KarhooConfiguration(identifier: identifier, referer: referer, organisationId: organisationId, isProduction: isProduction))
    }

    // KarhooUserService
    //

    @objc func register(_ registrationData: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooUserService.register(registrationData, resolve: resolve, reject: reject);
    }

    @objc func login(_ loginData: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooUserService.login(loginData, resolve: resolve, reject: reject);
    }

    @objc func logout(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooUserService.logout(resolve, reject: reject);
    }

    @objc func currentUser(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooUserService.currentUser(resolve, reject: reject);
    }

    // KarhooAuthService
    //

    @objc func loginWithToken(_ token: NSString, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooAuthService.loginWithToken(token, resolve: resolve, reject: reject);
    }

    // KarhooAddressService
    //

    @objc func placeSearch(_ placeSearchData: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooAddressService.placeSearch(placeSearchData, resolve: resolve, reject: reject);
    }

    @objc func locationInfo(_ locationInfoData: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooAddressService.locationInfo(locationInfoData, resolve: resolve, reject: reject);
    }

    // KarhooPaymentService
    // 

    @objc func getPaymentNonce(_ organisationId: String, paymentData: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {        
        return KarhooPaymentService.getPaymentNonce(organisationId, paymentData: paymentData, resolve: resolve, reject: reject);
    }

    // KarhooTripService
    //
    
    @objc func bookTrip(_ bookTripData: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
       return KarhooTripService.bookTrip(bookTripData, resolve: resolve, reject: reject);
    }

    @objc func cancellationFee(_ followCode: NSString, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooTripService.cancellationFee(followCode, resolve: resolve, reject: reject);
    }

    @objc func cancelTrip(_ followCode: NSString, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        return KarhooTripService.cancelTrip(followCode, resolve: resolve, reject: reject);
    }
}


